Add 'alphanumeric sort' commands

seb-bl opened the pull request:
Implements #4153 with the crate mentioned in the discussion of the issue.

This adds 2 new commands, `sortn` and `rsortn`, corresponding to the existing ones (`sort` and `rsort`), and they sort according to https://docs.rs/alphanumeric-sort/1.4.4/alphanumeric_sort/fn.compare_str.html

the-mikedavis commented:
It's a short (~100L) block of code within the library and the library includes a bunch of other functions I don't think will be used in the future so I think we should vendor this block https://github.com/magiclen/alphanumeric-sort/blob/bea0284409f68407dd25b432fd901b62f6e944e7/src/lib.rs#L95-L199

kirawi commented:
I think that https://github.com/Aloso/lexical-sort might be better since it claims to support unicode.

seb-bl commented:
I removed the dependency, and added a function to compare the strings.

When I copied the source, I realized the handling of long numbers was incorrect (it accumulates it in a `f64`, thus rounds it). I looked at the other crate, and it had a similar problem (accumulating the number in a `u64`).

So I implemented my own function. It orders numbers with leading zeroes like the crate `alphanumeric-sort`, but uses the slice containing the digits instead of a `f64` for the comparison

kirawi commented:
Just to note, I think that using a `u64` would've been okay. `len()` returns a `usize` which is equivalent to `u64` on 64bit systems.

kirawi commented on helix-term/src/commands/typed.rs:
```diff
@@ -1407,6 +1434,87 @@ fn sort_impl(
+    }
+}
+
+#[test]
```
This should be moved into a
```rs
#[cfg(test)]
mod tests {
}
```
block at the bottom of the file.

kirawi commented:
By the way, make sure to run `cargo xtask docgen` and commit the changes to make CI pass.

seb-bl commented:
Thanks for the hints 🙂

I think I have not explained the problem with the `u64` correctly. When encountering a number,`lexical-sort` parses the number into a `u64`, digit by digit, it multiplies the number by 10 and add the current digit ([at this line](https://github.com/Aloso/lexical-sort/blob/main/src/cmp.rs#L14-L15)). So for example "18446744073709551616" (=2^64) would cause an overflow. Numbers this long are not very common, but may still happen

leandrobbraga commented on helix-term/src/commands/typed.rs:
```diff
@@ -1370,13 +1370,38 @@ fn sort_reverse(
+    sort_impl(cx, args, true, true)
 }
 
 fn sort_impl(
     cx: &mut compositor::Context,
     _args: &[Cow<str>],
     reverse: bool,
+    alphanumeric: bool,
 ) -> anyhow::Result<()> {
```
As a suggestion you remove both `reverse` and `alphanumeric` and pass a single arguments called `cmp_fn` which is a `FnMut(&Tendril, &Tendril) -> Ordering` leaving to the caller to say how it want the ordering.

See: https://github.com/helix-editor/helix/pull/8436

the-mikedavis commented on helix-term/src/commands/typed.rs:
```diff
@@ -1388,9 +1413,11 @@ fn sort_impl(
-    fragments.sort_by(match reverse {
-        true => |a: &Tendril, b: &Tendril| b.cmp(a),
-        false => |a: &Tendril, b: &Tendril| a.cmp(b),
+    fragments.sort_by(match (reverse, alphanumeric) {
+        (true, false) => |a: &Tendril, b: &Tendril| b.cmp(a),
+        (false, false) => |a: &Tendril, b: &Tendril| a.cmp(b),
+        (true, true) => |a: &Tendril, b: &Tendril| numeric_cmp(b, a),
+        (false, true) => |a: &Tendril, b: &Tendril| numeric_cmp(a, b),
     });
```
I agree with @leandrobbraga - this block is a little dense and we can make the call sites for `sort_impl` a little cleaner by moving these closures to parameters of `sort_impl`.

So this function would replace the `reverse` parameter with a `compare_fn: impl FnMut(&Tendril, &Tendril) -> std::cmp::Ordering,` param and then this block can be just `fragments.sort_by(compare_fn);`.

We could also use this opportunity to drop the `args` parameter which is currently unused

kirawi commented:
I was looking into this kind of thing for a personal project and I don't think that alphanumeric-sort provides true lexicographic ordering. For that, we'd want the Unicode Collation algorithm. I'm currently a little less than halfway into my naive implementation of the algorithm that would be fully compliant so I may post a PR for it in the future.
